gordo_components/model/anomaly/diff.py

```diff
@@ -78,6 +79,112 @@ def fit(self, X: np.ndarray, y: np.ndarray):
         self.scaler.fit(y)  # Scaler is used for calculating errors in .anomaly()
         return self
 
+    def cross_validate(
+        self,
+        *,
+        X: Union[pd.DataFrame, np.ndarray],
+        y: Optional[Union[pd.DataFrame, np.ndarray]] = None,
+        cv=TimeSeriesSplit(n_splits=3),
+        **kwargs,
+    ):
+        """
+        Run cross validation on the model, and will update the model's threshold value based
+        on the cross validation folds
+
+        Parameters
+        ----------
+        X: Union[pd.DataFrame, np.ndarray]
+            Input data to the model
+        y: Optional[Union[pd.DataFrame, np.ndarray]]
+            Target data
+        kwargs: dict
+            Any additional kwargs to be passed to :func:`sklearn.model_selection.cross_validate`
+
+        Returns
+        -------
+        dict
+        """
+        # Depend on having the trained fold models
+        kwargs.update(dict(return_estimator=True, cv=cv))
+
+        X = X if not isinstance(X, pd.DataFrame) else X.values
+        y = y if not isinstance(y, pd.DataFrame) else y.values
+
+        # Base cv_output dict, which we'll supplement
+        cv_output = cross_validate(self, X=X, y=y, **kwargs)
+
+        thresholds = pd.DataFrame()
+
+        for i, ((test_idxs, _train_idxs), model) in enumerate(
+            zip(kwargs["cv"].split(X, y), cv_output["estimator"])
+        ):
+            y_pred = model.predict(X[test_idxs])
+            y_true = y[test_idxs]
+
+            diff = self._fold_thresholds(y_true=y_true, y_pred=y_pred, fold=i)
+
+            # Accumulate the rolling mins of diffs into common df
+            thresholds = thresholds.append(diff)
+
+        # And finally, get the mean of all the pre-calculated thresholds over the folds
+        self.thresholds_ = self._final_thresholds(thresholds=thresholds)
```

Comment:
This is a bit too subtle for me, that we have to run cross_validate on the same object before calling `anomaly` to have this variable set. Is there maybe some way we can make it a bit more explicit

Reply:
#false You only get thresholds if `cross_validate` method was called. It will still work (calling `.anomally()` that is) without calling `cross_validate` 

IMO, it is worth the trade off of an ugly wrapping of `cross_validate` and then re-initializing the same Model depending on some output. Implementing an optional `cross_validate` as an optional method, keeps things quite clean in the `build_model` function.

The requirement of _this_ model needing access to folds is exceptionally specific to _this_ model.

Reply:
otherwise, we can call `cross_validate` inside of `fit` if `thresholds_` is not defined?